src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Ssl.cs

```diff
@@ -0,0 +1,633 @@
+            }
+
+            Debug.Fail($"AppleCryptoNative_SslSetBreakOnClientAuth returned {result}");
+            throw new SslException();
+        }
+
+        internal static void SslSetCertificate(SafeSslHandle sslHandle, IntPtr[] certChainPtrs)
+        {
+            using (SafeCreateHandle cfCertRefs = CoreFoundation.CFArrayCreate(certChainPtrs, (UIntPtr)certChainPtrs.Length))
+            {
+                int osStatus = AppleCryptoNative_SslSetCertificate(sslHandle, cfCertRefs);
+
+                if (osStatus != 0)
+                {
+                    throw CreateExceptionForOSStatus(osStatus);
+                }
+            }
+        }
+
+        internal static void SslSetTargetName(SafeSslHandle sslHandle, string targetName)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(targetName));
+
+            int osStatus;
+            int cbTargetName = System.Text.Encoding.UTF8.GetByteCount(targetName);
+
+            int result = AppleCryptoNative_SslSetTargetName(sslHandle, targetName, cbTargetName, out osStatus);
+
+            if (result == 1)
+            {
+                return;
+            }
+
+            if (result == 0)
+            {
+                throw CreateExceptionForOSStatus(osStatus);
+            }
+
+            Debug.Fail($"AppleCryptoNative_SslSetTargetName returned {result}");
+            throw new SslException();
+        }
+
+        public static bool SslCheckHostnameMatch(SafeSslHandle handle, string hostName, DateTime notBefore)
+        {
+            int result;
+            // The IdnMapping converts Unicode input into the IDNA punycode sequence.
+            // It also does host case normalization.  The bypass logic would be something
+            // like "all characters being within [a-z0-9.-]+"
+            // Since it's not documented as being thread safe, create a new one each time.
+            //
+            // The SSL Policy (SecPolicyCreateSSL) has been verified as not inherently supporting
+            // IDNA as of macOS 10.12.1 (Sierra).  If it supports low-level IDNA at a later date,
+            // this code could be removed.
+            //
+            // It was verified as supporting case invariant match as of 10.12.1 (Sierra).
+            string matchName = new System.Globalization.IdnMapping().GetAscii(hostName);
+
+            using (SafeCFDateHandle cfNotBefore = CoreFoundation.CFDateCreate(notBefore))
+            using (SafeCreateHandle cfHostname = CoreFoundation.CFStringCreateWithCString(matchName))
+            {
+                result = AppleCryptoNative_SslIsHostnameMatch(handle, cfHostname, cfNotBefore);
+            }
+
+            switch (result)
+            {
+                case 0:
+                    return false;
+                case 1:
+                    return true;
+                default:
+                    Debug.Fail($"AppleCryptoNative_SslIsHostnameMatch returned {result}");
```

Comment:
Yeah, the shim function returns 1 on success, 0 when there's an OSStatus value to report, and any other value when the shim believes that the universe makes no sense.  Since we're (we believe) off the rails when a default value is returned, it's good to say we're dead in the water.

The actual non-normalized-bool return code only makes sense to the shim function that's being called, to track down where we went off the rails.  It's not stable codes, and doesn't have messages.